### PR TITLE
Reworked the application shell

### PR DESCRIPTION
**Pull request by spoenemann**
First step towards #1025.

I added a bottom panel to the application shell. The "Open Problems View" command now opens the view in that panel.

I also refactored and cleaned up some of the code in `shell.ts` and `sidepanel.css`.

### CODE REVIEW

**Comment by akosyakov** [comment]
Nice! Some things to take care:
- [ ] the bottom bar should look & feel like a tab bar in the main area, i.e. ability to close individual widgets, the context menu, black background
- [ ] there is empty space in a tab caption of sidebars, apparently for icons
- [ ] would be nice to open terminals in the bottom bar

**Comment by spoenemann** [comment]
* [ ] Add open commands for Files, Git, Extensions and Outline to the View menu.

**Comment by svenefftinge** [comment]
This will fix https://github.com/theia-ide/theia/issues/533

File: examples/browser/test/bottom-panel/bottom-panel.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
+import "webdriverio";
```
[the remark this answers is not in the document]
Comment (spoenemann) [reply examples/browser/test/bottom-panel/bottom-panel.ts]: &#9989;

File: packages/core/src/browser/shell.ts [hunk 2]
```diff
@@ -30,104 +32,109 @@ import { Saveable } from "./saveable";
...
+    type: 'sidebar',
+    expandedWidgets?: Widget[];
     widgets?: Widget[];
 }
 
-export interface DockLayoutData extends DockPanel.ILayoutConfig {
-    activeWidgets?: Widget[]
-}
+export const MAIN_AREA_TABBAR_CONTEXT_MENU: MenuPath = ['main-area-tabbar-context-menu'];
+export const SIDE_AREA_TABBAR_CONTEXT_MENU: MenuPath = ['side-area-tabbar-context-menu'];
```
Comment (svenefftinge) [inline packages/core/src/browser/shell.ts]: we should have one context menu for the tabs and remove certain commands by implementing `isVisible`

File: packages/navigator/src/browser/navigator-frontend-module.ts [hunk 3]
```diff
@@ -13,15 +13,23 @@ import { NavigatorMenuContribution } from './navigator-menu';
...
 
 export default new ContainerModule(bind => {
     bind(FileNavigatorContribution).toSelf().inSingletonScope();
     bind(FrontendApplicationContribution).toDynamicValue(c => c.container.get(FileNavigatorContribution));
-    bind(WidgetFactory).toDynamicValue(c => c.container.get(FileNavigatorContribution));
+    bind(CommandContribution).toDynamicValue(c => c.container.get(FileNavigatorContribution));
+    bind(KeybindingContribution).toDynamicValue(c => c.container.get(FileNavigatorContribution));
     bind(MenuContribution).to(NavigatorMenuContribution).inSingletonScope();
+
     bind(FileNavigatorWidget).toDynamicValue(ctx =>
```
Comment (svenefftinge) [inline packages/navigator/src/browser/navigator-frontend-module.ts]: could you remove this binding, so it is not exposed and accidentally used?

File: packages/navigator/src/browser/navigator-container.ts [hunk 4]
```diff
@@ -41,5 +45,25 @@ export function createFileNavigatorContainer(parent: interfaces.Container): Cont
 }
 
 export function createFileNavigatorWidget(parent: interfaces.Container): FileNavigatorWidget {
-    return createFileNavigatorContainer(parent).get(FileNavigatorWidget);
+    const fileNavigator = createFileNavigatorContainer(parent).get(FileNavigatorWidget);
+
+    const selectionService = parent.get(SelectionService);
```
Comment (akosyakov) [inline packages/navigator/src/browser/navigator-container.ts]: Could we move this logic to `FileNavigatorWidget` constructor?

File: packages/navigator/src/browser/navigator-frontend-module.ts (hunk 3 above)
Comment (spoenemann) [reply packages/navigator/src/browser/navigator-frontend-module.ts]: There are bindings for other widgets, too: `problem-frontend-module.ts`, `git-frontend-module.ts`, `extension-frontend-module.ts`. What about those?

**Comment by svenefftinge** [comment]
When I first open it (or use reset layout) it shows an empty expanded bottom panel:

screen shot 2018-01-10 at 16 24 29

It should be collapsed and when empty not visible at all.

**Comment by svenefftinge** [comment]
The commands should be renamed to `Toggle <View>` and they should close the view, if the view is currently focussed, otherwise open and focus.

Terminals are not singleton widgets and therefore don't really make sense in the 'view' space. I'd suggest to move the command back to `File` and call it `New Terminal`.

For the layout we should just change it so that the side panels get the full height and the bottom and main area the rest.

File: packages/extension-manager/src/browser/extension-contribution.ts [hunk 5]
```diff
@@ -34,11 +49,4 @@ export class ExtensionContribution implements FrontendApplicationContribution {
         });
     }
 
-    async initializeLayout(app: FrontendApplication): Promise<void> {
-        const extensionWidget = await this.widgetManager.getOrCreateWidget('extensions');
-        app.shell.addToLeftArea(extensionWidget, {
-            rank: 300
```
Comment (akosyakov) [inline packages/extension-manager/src/browser/extension-contribution.ts]: What about rank how the default order could be defined? the same for Git

File: packages/navigator/src/browser/navigator-widget.ts [hunk 6]
```diff
@@ -30,12 +33,33 @@ export class FileNavigatorWidget extends FileTreeWidget {
...
+        @inject(LabelProvider) labelProvider: LabelProvider
     ) {
         super(props, model, contextMenuRenderer);
         this.id = FILE_NAVIGATOR_ID;
         this.title.label = LABEL;
         this.addClass(CLASS);
+        this.initialize(selectionService, workspaceService, labelProvider);
+    }
+
+    protected initialize(selectionService: SelectionService, workspaceService: WorkspaceService, labelProvider: LabelProvider): void {
```
Comment (akosyakov) [inline packages/navigator/src/browser/navigator-widget.ts]: args can be properties instead?

**Comment by akosyakov** [comment]
I got the following error trying it out:
```
Uncaught (in promise) TypeError: Cannot set property 'parent' of undefined
    at SideBarHandler.addWidget (shell.ts:988)
    at SideBarHandler.setLayoutData (shell.ts:937)
    at ApplicationShell.setLayoutData (shell.ts:313)
    at shell-layout-restorer.ts:71
    at <anonymous>
```

**Comment by akosyakov** [comment]
Probably out of the scope: keybindings are displayed only in the selection menu, should not they be displayed in all menus? e.g in `View` it will be handy

**Comment by spoenemann** [comment]
> Probably out of the scope: keybindings are displayed only in the selection menu, should not they be displayed in all menus?

#490

**Comment by svenefftinge** [comment]
Looks like the test failures are 'real'

File: packages/core/src/browser/shell/view-contribution.ts [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+        }
+    }
+
+    get widget(): Promise<T> {
+        return this.widgetManager.getOrCreateWidget<T>(this.options.widgetId);
+    }
+
+    async openView(args: Partial<OpenViewArguments> = {}): Promise<T> {
+        const widget = await this.widget;
+        const tabBar = this.shell.gettTabBarFor(widget);
```
Comment (akosyakov) [inline packages/core/src/browser/shell/view-contribution.ts]: typo: `gettTabBarFor` ?

**Comment by spoenemann** [comment]
In the course of reacting to review comments, I have refactored and cleaned up a lot of code, so this PR has now become much more than just adding a bottom panel. It's a complete rework of the application shell and its API.

Note that I have touched almost all public methods of `ApplicationShell`, so this PR is a breaking change.

File: packages/terminal/src/browser/terminal-frontend-contribution.ts [hunk 8]
```diff
@@ -84,10 +84,9 @@ export class TerminalFrontendContribution implements CommandContribution, MenuCo
         /* Register ctrl + k (the passed Key) as a passthrough command in the
            context of the terminal.  */
         const regCtrl = (k: Key) => {
-            const modifier = isOSX ? Modifier.M4 : Modifier.M1;
             keybindings.registerKeybindings({
                 commandId: KeybindingRegistry.PASSTHROUGH_PSEUDO_COMMAND,
-                keyCode: KeyCode.createKeyCode({ first: k, modifiers: [modifier] }),
+                keyCode: KeyCode.createKeyCode({ first: k, modifiers: [Modifier.CTRL] }),
```
Comment (hexa00) [inline packages/terminal/src/browser/terminal-frontend-contribution.ts]: why is that changed?

Comment (spoenemann) [reply packages/terminal/src/browser/terminal-frontend-contribution.ts]: I extracted the check `isOSX ? Modifier.M4 : Modifier.M1` to a new constant that can (should) be used where CTRL is needed independently of the platform. The behavior is unchanged.

**Comment by hexa00** [comment]
Looks great in general! 

It may need quite a lot of manual testing with all it changes, I'll try to play around with it today.

File: packages/terminal/src/browser/terminal-frontend-contribution.ts (hunk 8 above)
Comment (hexa00) [reply packages/terminal/src/browser/terminal-frontend-contribution.ts]: OK indeed nice :)

**Comment by spoenemann** [comment]
There are new key bindings:

* M1 + Shift+ E: toggle Files
* Ctrl + Shift + G: toggle Git
* M1 + Shift + X: toggle Extensions
* Ctrl + Shift + O: toggle Outline
* M1 + Shift + M: toggle Problems
* Alt + C: collapse side panel
* Alt + Shift + C: collapse all side panels

**Comment by hexa00** [comment]
Feel free to add that to the commit log :)

**Comment by AlexTugarev** [comment]
@spoenemann just FYI, the branch needs an update. `@theia/output` was added and needs to learn new API.

**Comment by AlexTugarev** [comment]
@spoenemann, I'd like to propose adding optional options to main panel, similar to what I've tried out for markdown preview extension, cf. https://github.com/theia-ide/theia/pull/1051/commits/0cdaa1d5c90c79adb73672d093b3c9c805c081cc and demo: https://github.com/theia-ide/theia/pull/1051#issue-288519297

File: packages/core/src/browser/shell/view-contribution.ts [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+        const widget = await this.widget;
+        const tabBar = this.shell.getTabBarFor(widget);
+        if (!tabBar) {
+            // The widget is not attached yet, so add it to the shell
+            const widgetArgs: OpenViewArguments = {
+                ...this.options.defaultWidgetOptions,
+                ...args
+            };
+            this.shell.addWidget(widget, widgetArgs);
+        } else if (tabBar.currentTitle === widget.title && args.toggle) {
```
Comment (akosyakov) [inline packages/core/src/browser/shell/view-contribution.ts]: `widget` has `isVisible` attribute, could it be used here?

Comment (akosyakov) [reply packages/core/src/browser/shell/view-contribution.ts]: or `this.shell.currentWidget === widget` it should mean that a widget is attached, visible and in focus

Comment (spoenemann) [reply packages/core/src/browser/shell/view-contribution.ts]: The widget should be closed when it is already visible in order to realize the toggle behavior. The tab bars indicate which widgets are visible.

**Comment by spoenemann** [comment]
> `@theia/output` was added and needs to learn new API.

&#9989;

> I'd like to propose adding optional options to main panel, similar to what I've tried out for markdown preview extension

&#9989;

Also fixed an issue regarding shell layout storage where the order of widgets was not consistent after reloading.

**Comment by svenefftinge** [comment]
Nice!, please merge.
